File: dictionaries-and-hashmaps/02-SherlockAndAnagrams.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the sherlockAndAnagrams function below.
def sherlockAndAnagrams(s):
    anagramCount = 0
    for length in range (1, len(s)):  # iterate over all substrings of length 1 to n-1
        # get all possible sub strings
        subs = {}
        for i in range (0, len(s)-length+1):
            mainSubstring = s[i:i+length]
            # print("main: "+mainSubstring)
            for j in range (i+1, len(s)-length+1):
                nextSubstring = s[j:j+length]
                # sort string ''.join(sorted(a))
                
                # print(nextSubstring)
                
                if sorted(mainSubstring) == sorted(nextSubstring):  
                    anagramCount += 1
        
        # if no single letter matches found, anagram not possible
        if length == 1 and anagramCount == 0:
            break        
            
        # look for matches
    return anagramCount

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    q = int(input())

    for q_itr in range(q):
        s = input()

        result = sherlockAndAnagrams(s)

        fptr.write(str(result) + '\n')

    fptr.close()

File: dictionaries-and-hashmaps/02-SherlockAndAnagrams.cpp
#include <bits/stdc++.h>
#include <string>

using namespace std;

// Complete the sherlockAndAnagrams function below.
int sherlockAndAnagrams(string s) {
    int anagramCount = 0;
    //  iterate over all substrings of length 1 to n-1
    for (int length = 1; length < s.length(); length++) {
        for (int i = 0; i < s.length() - length + 1; i++) {
            string mainSubstring = s.substr(i,length);
            // cout << "main:" << mainSubstring << endl;
            for (int j = i+1; j < s.length() - length + 1; j++) {
                string nextSubstring = s.substr(j,length);
                
                // cout << nextSubstring << endl;
                
                string sortMainSubstring = mainSubstring;
                string sortNextSubstring = nextSubstring;
                
                std::sort(sortMainSubstring.begin(), sortMainSubstring.end());
                std::sort(sortNextSubstring.begin(), sortNextSubstring.end());
                if (sortMainSubstring == sortNextSubstring ) {
                    anagramCount += 1;
                }
            }                    
        }
        
        // if no single letter matches found, anagram not possible
        if (length == 1 && anagramCount == 0) {
            break;
        }
    }
    // look for matches
    return anagramCount;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int q;
    cin >> q;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    for (int q_itr = 0; q_itr < q; q_itr++) {
        string s;
        getline(cin, s);

        int result = sherlockAndAnagrams(s);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}

